### ISSUE TITLE
Support service application type

### ISSUE THREAD

**Issue by rwalczyna** [flutter-tizen/flutter-tizen#94]
The goal is to add support for Tizen Service Application that could run Dart code in the background without UI and communicate with other Flutter applications.

Main capabilities:

* Performing complex / time consuming headless computation in separate process
* Starting service application from different Flutter application
* Sending commands from different Flutter application to the service
* Communication between service and other Flutter applications

TODO:

Changes in tools:

- [ ] [engine](https://github.com/flutter-tizen/engine) - adding support for headless application (IN PROGRESS: flutter-tizen/engine#66)
- [ ] [flutter-tizen](https://github.com/flutter-tizen/flutter-tizen/) - creation and running tizen service application (preparing templates)

Changes in application runner:
- [ ] Tizen service application with service_app_main instead of ui_app_main
- [ ] Starting service application and receiving simple commands (service_app_control_cb)
- [ ] Bi-directional communication using Message Port (probably creation of flutter plugin)

Related native solutions: [Service Application](https://docs.tizen.org/application/native/guides/applications/service-app/), [Message Port](https://docs.tizen.org/application/native/guides/app-management/message-port/), [Application Control](https://docs.tizen.org/application/web/guides/app-management/app-controls/)

**Comment by swift-kim** [comment]
I have a few questions:

1. https://docs.tizen.org/application/native/guides/applications/service-app/#attribute says the `auto-restart` and `on-boot` features are no longer supported on all Tizen devices. Should a service app always be launched manually from the command line? I also read https://docs.tizen.org/application/native/guides/applications/efl-ui-app/#background-categories but it's still unclear. (Should a service app be always packaged with a UI app to operate properly?)
2. https://docs.tizen.org/application/native/guides/applications/service-app/#application-states says a service app can go into the suspended state. Do you know any details about it, like how the suspended app can wake up and resume the execution?

**Comment by rwalczyna** [comment]
1. These two features are not available on wearable - due to battery consumption optimization. On other devices it should work properly. I am testing it on TM1 and it works as expected. It is not documented, but I think that on wearable developer should use workaround to be sure that application runs properly (setting alarm to reopen app or using companion UI application).
Regarding `background-category` it looks like without it application is blocked by system after a while (to save resources).
2. I couldn't find any documentation about suspended state for service application. Only some advice is available for UI application [here](https://docs.tizen.org/application/native/api/iot-headed/6.0/group__CAPI__APPLICATION__MODULE.html#CAPI_APPLICATION_MODULE_STATES_TRANSITIONS). But as service app does not support `resume` event I think it is not applicable.

To sum up, I think that event `suspended` on service application can occur only when there are not enough resources and resource manager decides to limit an application, but there is no official information about it. Application should `resume` or `start again` on `app_control` request, so UI application should be always capable of waking up service application.

**Comment by rwalczyna** [comment]
@swift-kim, I have a question regarding communication plugin. I have prepared the implementation of Message Port plugin - [here](https://github.com/rwalczyna/plugins/tree/messageport_devel/packages/messageport). 
I have created file structure similar to other plugins, as I couldn't find any guide for new plugin creation process.
I checked [plugins](https://github.com/flutter-tizen/plugins/) repository and I noticed that some of newly implemented plugins has their own branch - `plugin_devel`. Should I create PR to `master` branch or to `messageport_devel` branch?

**Comment by swift-kim** [comment]
> I couldn't find any guide for new plugin creation process.

I have to write some guide as I recently told in https://github.com/flutter-tizen/plugins/issues/109 - maybe in this week. You may also find a pre-release checklist in the f-project/plugins (internal repo) wiki.

> Should I create PR to `master` branch or to `messageport_devel` branch?

Not generally. The branches were created for plugins like webview_flutter_tizen or camera_tizen which were complex enough and had to be implemented through multiple PRs. You can submit your PR to the master branch if the implementation is almost done.
